myip/myip.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// TaobaoIPURL is the API endpoint to get external IP address from Taobao.
+const TaobaoIPURL = "http://ip.taobao.com/service/getIpInfo.php?ip=myip"
+
+func init() {
+	TaobaoRequest, _ = http.NewRequest(http.MethodGet, TaobaoIPURL, nil)
+}
+
+type MyIP struct {
+	sync.RWMutex
+	ip  net.IP
+	api Interface
+}
+
+// Interface defines the interface to get external IP address of running code.
+type Interface interface {
+	IP() (net.IP, error)
+}
+
+// Taobao implements Interface using Taobao's API
+type Taobao struct {
+	Interface
+
+	Client *http.Client
+}
+
+func (m *MyIP) refreshFromTaobaoIP() error {
+	ip, err := m.api.IP()
+	if err != nil {
+		return err
+	}
+	m.SetIP(ip)
+	return nil
+}
+
+func (m *MyIP) GetIP() net.IP {
+	m.RLock()
+	defer m.RUnlock()
+	return m.ip
+}
+
+func (m *MyIP) SetIP(ip net.IP) {
+	m.Lock()
+	m.ip = ip
+	m.Unlock()
+}
+
+// New return a new MyIP instance to get external IP address from given API.
+func New(api Interface) *MyIP {
+	return &MyIP{api: api}
+}
+
+func (m *MyIP) StartTaobaoIPLoop(cb func(oldIP, newIP net.IP)) {
```

Comment:
Ah, yes, `ifconfig.co` support IPv6 too, which is more reliable than running one's own App Engine instance. Will work on additional code that uses `ifconfig.co`